🔴 [VLM] Add base model without head

zucchini-nlp opened the pull request:
# What does this PR do?

Stage one of vLLM with transformers backend support for vision LLMs. As discussed internally, we don't want to break existing models, so we sacrifice readability and duplicate code

The PR adds base models for all model where missing. The base model is supposed to be same as in LLMs, everything except for the head. This allows us to make modeling more aligned with vLLM and to have a standard API for multimodal generation models. Will be super helpful in the long run, for example for `AutoToAny` mapping

Next stages for modeling to help vLLM and TGI:
- Add new attention interface where still absent
- Qwen2 config workaround without breaking
- Processor standardization sprint
    - Add attributes for image_token_id/image_token if missing
    - Return `mm-token-type-ids` if requested with to indicate where image/video/audio placeholder are
    - Add helper `get_num_of_image_tokens` for all processor, which returns placeholder length given image
    - Explore what else missing for processors
- Finalize and merge PR on vLLM repo, check correctness for different models


Fixes https://github.com/huggingface/transformers/issues/36940.

zucchini-nlp committed:
i guessreverted all CdGen classes

zucchini-nlp committed:
style

zucchini-nlp committed:
llava onevision

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37033). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp commented on src/transformers/models/qwen2_vl/modeling_qwen2_vl.py:
```diff
@@ -1044,7 +1081,7 @@ def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor) -> torch.
     QWEN2VL_START_DOCSTRING,
 )
-class Qwen2VLModel(Qwen2VLPreTrainedModel):
+class Qwen2VLTextModel(Qwen2VLPreTrainedModel):
```
this might seem breaking. I terms of what we expect at input and what we return from new `Qwen2VL` module, nothing changes. We can accept more args now but text-only still fine and returns same hidden states

zucchini-nlp committed:
fix copies

zucchini-nlp commented:
not ready yet, marking ready to get CI run

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp committed:
fix some tests

zucchini-nlp committed:
some more tests

zucchini-nlp committed:
dump

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp committed:
skip these

zucchini-nlp committed:
nevermind, i am dumb

zucchini-nlp committed:
revert fix not needed

zucchini-nlp commented:
@ArthurZucker  ready for review

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp committed:
fixup

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

zucchini-nlp commented:
Thanks @jacob-danner ! These are probably copied from the very first model, from which we copy all other. In this case it might be LLaVA, I will check it

zucchini-nlp committed:
fixup

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm-base-models

zucchini-nlp committed:
another fixup

zucchini-nlp committed:
more fixup to make ci finally happy

pySilver commented:
@zucchini-nlp Sorry to bother you. Is this going to be merged any time soon? I'm asking mainly because of your comment here https://github.com/huggingface/transformers/issues/36755#issuecomment-2757756750 Sadly I cannot be any help in this field :(

zucchini-nlp commented:
Sorry yeah, the core maintainer is off and the review got stale. cc @Cyrilvallez if you have bandwidth to review it :)

I will rebase main again soon

zucchini-nlp committed:
merge main

zucchini-nlp committed:
fixup after rebasing

zucchini-nlp committed:
fix qwen tests

zucchini-nlp committed:
add internVL + typos here and there

zucchini-nlp committed:
image token index -> id

zucchini-nlp committed:
style

zucchini-nlp committed:
fix init weights

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm-base-models

zucchini-nlp committed:
Merge branch 'main' into vlm-base-models

Cyrilvallez commented on src/transformers/models/aya_vision/modeling_aya_vision.py:
```diff
@@ -117,7 +117,7 @@ class AyaVisionPreTrainedModel(PreTrainedModel):
     base_model_prefix = "model"
     supports_gradient_checkpointing = True
-    _no_split_modules = ["AyaVisionVisionAttention"]
+    _no_split_modules = ["LlamaDecoderLayer"]
```
Here both the old one and the new one look like they are wrong no? Or is it because the main backbone is Llama?

zucchini-nlp replied:
yeah, `AyaVisionVisionAttention` seemed incorrect. But now I remembered that we would dispatch `_no_split_modules` from module children so we can remove it probably. I will double check, in case it doesn't work we need a new PR for that

ArthurZucker replied:
yes children are taken care of normallY!

Cyrilvallez commented on src/transformers/models/blip_2/modeling_blip_2.py:
```diff
@@ -1532,16 +1532,11 @@ def __init__(self, config: Blip2Config):
-            language_model = AutoModelForCausalLM.from_config(config.text_config)
-        else:
-            language_model = AutoModelForSeq2SeqLM.from_config(config.text_config)
+        self.language_model = AutoModel.from_config(config.text_config)
```
This is a bit breaking here, but not sure we have a better way? 🥲

zucchini-nlp replied:
I know 😢  I just started working on other parts for vLLM and in every case BLIP makes an exception and breaks BC. So now I'm thinking to not support BLIP at all, and focus on recent models

will remove this, and add a comment that this way is discouraged for those users who copy from BLIP

ArthurZucker replied:
this sounds better! lets explicitly not support blip

Cyrilvallez commented on src/transformers/models/internvl/modeling_internvl.py:
```diff
@@ -608,14 +608,14 @@ def forward(
     base_model_prefix = "model"
     supports_gradient_checkpointing = True
-    _no_split_modules = ["InternVLVisionAttention"]
+    _no_split_modules = ["LlamaDecoderLayer"]
```
same comment here, is it because of a Llama backbone?

Cyrilvallez commented on src/transformers/modeling_utils.py:
```diff
@@ -1757,6 +1757,8 @@ class PreTrainedModel(nn.Module, ModuleUtilsMixin, PushToHubMixin, PeftAdapterMi
     main_input_name = "input_ids"
     model_tags = None
 
+    _key_mapping = None  # used for BC support in VLMs, not meant to be used by new models
+
```
Here I think I would prefer not to have the default at all

Cyrilvallez commented on src/transformers/modeling_utils.py:
```diff
@@ -4027,7 +4029,7 @@ def from_pretrained(
         tp_plan = kwargs.pop("tp_plan", None)
         tp_size = kwargs.pop("tp_size", None)
-        key_mapping = kwargs.pop("key_mapping", None)
+        key_mapping = kwargs.pop("key_mapping", cls._key_mapping)
```
And so only use it here if it exists
```suggestion
        key_mapping = kwargs.pop("key_mapping", getattr(cls, "._key_mapping", None))
```

But just a bit worried if it could become an issue when saving then reloading?

Cyrilvallez reviewed (commented):
Big big PR 🚀🤗
I'm mostly worried about the use of the automatic `key_mapping` here, as we don't modify the keys when saving, I think it can become an issue when saving then reloading no?

ArthurZucker commented on src/transformers/models/aria/modular_aria.py:
```diff
@@ -1378,9 +1397,136 @@ class AriaCausalLMOutputWithPast(LlavaCausalLMOutputWithPast):
 """
 
 
+class AriaModel(LlavaModel):
```
Given this, I wonder if `AriaForConditionalGeneration` can't inherit from `AriaModel` just for all the static functions that are the same? but its a nit

zucchini-nlp replied:
thought same thing but it wasn't as straightforward as it seemed. Simply inheriting `AriaForConditionalGeneration(AriaModel, GenerationMixin)` slows down things because we initialize model two times via super. We can technically make a helper class mixin and use it in base model and task model, but it's not aligned with transformers style

I realize we are duplicating some functions for image processing, but I think it is the easiest way and we just need it to force new models follow this standard (and help vLLM ofc). To make this a bit easier to maintain when fixing bugs, I will add `copy from` statements for functions like `get_image_features`

ArthurZucker replied:
No I mean in the modular only, just to have less code! not in the modelling!

ArthurZucker replied:
don't worry 🤗

ArthurZucker reviewed (commented):
In general LGMT, may main comment is that modular should be helping you here, by allowing you to set a function to be the same in another class! 
Like 
```AriaForCausalLM.xxxx = AriaModel.xxxx```

This will help maintenance!

ArthurZucker commented:
we could also have the ForXX class call the Model's function in the full modeling file, to keep bc but have a single place where we define the function

zucchini-nlp committed:
revert blip-2 not supported

zucchini-nlp committed:
address comments

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm-base-models

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm-base-models

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
revert blip2 test file as well

zucchini-nlp commented:
@ArthurZucker @Cyrilvallez addressed comments, ready for review! The last question is do we want these new base classes to be loadable or not? Note, it is not a requirement for vLLM but rather something users might need. If not, I will remove `key_mapping` and I hope we're ready to merge

zucchini-nlp committed:
as discussed internally, revert back CdGen models

zucchini-nlp committed:
fix some tests

zucchini-nlp committed:
fix more tests for compile

zucchini-nlp committed:
CI red

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
enumerate explicitly allowed models

ArthurZucker commented on src/transformers/models/gemma3/modeling_gemma3.py:
```diff
@@ -1066,20 +1100,19 @@ def forward(self, vision_outputs: torch.Tensor):
 )
-class Gemma3ForConditionalGeneration(Gemma3PreTrainedModel, GenerationMixin):
+class Gemma3Model(Gemma3PreTrainedModel):
+    _key_mapping = {"language_model.model": "language_model"}
```
as this appears in the model class, I would be happy to have a more meaningful name here! 
`_on_the_fly_conversion_mapping` or something like this!

ArthurZucker reviewed (commented):
The only thing missing for me is the ability to save original checkpoints (so a flag, as the `key_mapping` is only taken into account for the `loading` and not for the `saving` of the checkpoint!

zucchini-nlp committed:
address comments

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -3493,6 +3493,21 @@ def save_pretrained(
+            reverse_key_mapping = {v: k for k, v in self._checkpoint_conversion_mapping.items()}
+
+            original_state_dict = {}
+            for key, value in state_dict.items():
+                for pattern, replacement in reverse_key_mapping.items():
+                    replacement = replacement.lstrip("^")  # strip off un-needed chars and patterns
+                    replacement = re.sub(r"\(.*?\)", "", pattern)
+                    new_key, n_replace = re.subn(pattern, replacement, key)
+                    # Early exit of the loop
+                    if n_replace > 0:
+                        break
```
I don't mind iterating over the k, value, but the replacement can be easier / faster. you don't have to itrate over them! (inneficient)
You can: 
- pipe the patterns with `|` beforhand

zucchini-nlp replied:
We could do if the replacement was same across patterns. When we need to replace depending on which pattern is matched, it gets harder. I left as is for now

ArthurZucker reviewed (approved):
Thanks!

zucchini-nlp committed:
fix tests

zucchini-nlp committed:
fixup

zucchini-nlp committed:
merge main

zucchini-nlp committed:
style again

zucchini-nlp committed:
add tests for new model class

zucchini-nlp committed:
another fixup ( x _ x )

zucchini-nlp committed:
[fixup] unused attributes can be removed post-deprecation

ManuelFay commented on src/transformers/modeling_utils.py:
```diff
@@ -4071,7 +4110,13 @@ def from_pretrained(
-        key_mapping = kwargs.pop("key_mapping", None)
+
+        # Load models with hardcoded key mapping on class for VLMs only,  to keep BC and standardize model
+        if any(allowed_name in cls.__name__.lower() for allowed_name in VLMS):
```
This is quite brittle and breaks adapters (in peft). How would you go about this?
I'm thinking we can propagate the key_mapping to the peft library in the from pretrained function ?

ManuelFay replied:
Since a lot of people (including me) use adapters with VLMs, that's quite a big breaking change

ManuelFay reviewed (commented):
@zucchini-nlp To detail this a bit more, if loading with adapters (assuming the class name matches the hardcoded list (quite an assumption), or that the user specified a key_mapping, the adapters will not benefit from the key_mapping that the base model uses. 

Should we propagate this down ?

ManuelFay commented:
line 229, transformers.integrations.peft.py

```python
# We need to pre-process the state dict to remove unneeded prefixes - for backward compatibility
processed_adapter_state_dict = {}
prefix = "base_model.model."
for key, value in adapter_state_dict.items():
    if key.startswith(prefix):
        new_key = key[len(prefix) :]
    else:
        new_key = key

    import re
    if key_mapping:
        for pattern, replacement in key_mapping.items():
            new_key, n_replace = re.subn(pattern, replacement, new_key)
            # Early exit of the loop
            if n_replace > 0:
                has_changed = True
                break
    processed_adapter_state_dict[new_key] = value
```

line 152 is needed as well

` key_mapping = adapter_kwargs.pop("key_mapping", None) if adapter_kwargs is not None else None`


and probably something for saving ?
